src/combinatorics/set_family/compression/uv.lean

```diff
@@ -0,0 +1,265 @@
+disjoint and `v ≤ a`. In some sense, it's moving `a` from `v` to `u`.
+
+UV-compressions are immensely useful to prove the Kruskal-Katona theorem. The idea is that
+compressing a set family might decrease the size of its shadow, so iterated compressions hopefully
+minimise the shadow.
+
+## Main declarations
+
+* `uv.compress`: `compress u v a` is `a` compressed along `u` and `v`.
+* `uv.compress_motion`: `compress_motion u v s` is the part of `s` that can move along the
+  compression. Precisely, it's the compression of the elements of `s` whose compression is not
+  already in `s`.
+* `uv.compress_remains`: `compress_remains u v s` is the part of `s` that can't move along the
+  compression. Precisely, it's the element whose compression is already in `s`.
+* `uv.compression`: `compression u v s` is the compression of the set family `s` along `u` and `v`.
+  It is the union of `compress_motion u v s` and `compress_remains u v s`. This way of splitting
+  into what moves and what does not ensures the compression doesn't squash the set family, which is
+  proved by `uv.card_compress`.
+
+## Notation
+
+`𝓒` (typed with `\MCC`) is notation for `uv.compression` in locale `finset_family`.
+
+## Notes
+
+Even though our emphasis is on `finset α`, we define UV-compressions more generally in a generalized
+boolean algebra, so that one can use it for `set α`.
+
+## TODO
+
+Prove that compressing reduces the size of shadow. This result and some more already exist on the
+branch `combinatorics`.
+
+## References
+
+* https://github.com/b-mehta/maths-notes/blob/master/iii/mich/combinatorics.pdf
+
+## Tags
+
+compression, UV-compression, shadow
+-/
+
+open finset
+
+variable {α : Type*}
+
+-- The namespace here is useful to distinguish from other compressions.
+namespace uv
+
+/-! ### UV-compression in generalized boolean algebras -/
+
+section generalized_boolean_algebra
+variables [generalized_boolean_algebra α] [decidable_rel (@disjoint α _)]
+  [decidable_rel ((≤) : α → α → Prop)] {s : finset α} {u v a b : α}
+
+/-- To UV-compress `a`, if it doesn't touch `U` and does contain `V`, we remove `V` and
+put `U` in. We'll only really use this when `|U| = |V|` and `U ∩ V = ∅`. -/
+def compress (u v a : α) : α := if disjoint u a ∧ v ≤ a then (a ⊔ u) \ v else a
+
+/-- Part of the compressed family, where we move the sets whose compression is not there. -/
+def compress_motion (u v : α) (s : finset α) :=
+(s.filter (λ a, compress u v a ∉ s)).image (λ a, compress u v a)
+
+/-- Part of the compressed family, where we keep sets whose compression is already present. -/
+def compress_remains (u v : α) (s : finset α) := s.filter (λ a, compress u v a ∈ s)
+
+/-- To UV-compress a set family, we compress each of its elements, except that we don't want to
+reduce the cardinality, so we keep all elements whose compression is already present. -/
+def compression (u v : α) (s : finset α) := compress_motion u v s ∪ compress_remains u v s
```

Comment:
What do people think of inlining the definitions of `compress_motion` and `compress_remains`? They aren't used besides API making for `compression`. cc @b-mehta

Reply:
It's not such a clear win. Most of the gained simplicity comes from changing `compression` from
```
(s.filter (λ a, compress u v a ∉ s)).image (λ a, compress u v a)
  ∪ s.filter (λ a, compress u v a ∈ s)
```
to
```
s.filter (λ a, compress u v a ∈ s) ∪ (s.image $ compress u v).filter (λ a, a ∉ s)
```
by commuting `filter` and `image`.